add support for unfurl_links + better logging

dbertram opened the pull request:
- add support for optional `unfurl_links` argument (defaults to `true`, but can be used to disable link unfurling when desired)
- use `--fail-with-body` so HTTP status of >= 400 triggers an error exit code
- use long form curl args in general for better readability/comprehensibility
- break up curl args over multiple lines to improve readability
- log raw response before processing it
- fix how this action sets its `ts` output so that it actually works!
- adds a GH workflow to this repo to enable testing changes in PRs _before_ they're merged and without having to setup private, throw-away repos that use your branch of this GH action for testing

![](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExcGQ2Y29vanUwMWVnbmloNmd6cnUyaDhhMnJ4MWU1MDh4czBkNG93ZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/VXCPgZwEP7f1e/giphy.webp)

dbertram committed:
refactor how curl is called and errors are captured

- use --fail-with-body so HTTP status of >= 400 triggers an error exit code
- use long form curl args in general for better readability/comprehensibility
- break up curl args over multiple lines to improve readability
- log raw response before processing it
- tweak multi-line GITHUB_OUTPUT capture to more closely match doc examples

dbertram committed:
add support for optional unfurl_links argument

dbertram commented on action.yml:
```diff
@@ -23,11 +27,17 @@ runs:
   steps:
     - run: |
+        response=$(curl --fail-with-body --silent --show-error \
+          --request POST \
+          --header "Authorization: Bearer ${{ env.SLACK_BOT_TOKEN || inputs.token }}" \
+          --header "Content-Type: application/json; charset=utf-8" \
+          --url https://slack.com/api/chat.postMessage \
+          --data "{\"channel\": \"${{ inputs.channel }}\", \"thread_ts\": \"${{ inputs.thread_ts }}\", \"unfurl_links\": ${{ inputs.unfurl_links }}, \"text\": \"${{ inputs.text }}\"}" \
+        )
+        echo "Slack API response:\n$response"
         {
-          echo "ts=<<EOF"
-          curl --silent --show-error -X POST -H "Authorization: Bearer ${{ env.SLACK_BOT_TOKEN || inputs.token }}" -H "Content-Type: application/json; charset=utf-8" --url https://slack.com/api/chat.postMessage \
-            -d "{\"channel\": \"${{ inputs.channel }}\", \"thread_ts\":\"${{ inputs.thread_ts }}\", \"text\":\"${{ inputs.text }}\"}" \
-            | jq --raw-output '.ts'
+          echo "ts<<EOF"
+          echo "$response" | jq --raw-output '.ts'
```
Something about how I was trying to output the `ts` output variable wasn't playing nicely with what GH was expecting:

![image](https://github.com/user-attachments/assets/be615ece-a3ac-4a43-bd51-c8687c2cdba4)

I was looking more closely at their examples for how to do multi-line output and noticed I had added the equal sign in `ts=<<EOF` that wasn't in their example: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#multiline-strings

...but I'm not sure how test test this assumption before merging this PR. 😟

bradymholt replied:
You could create a workflow in a throw-away repo that references this branch like this:

```
- name: "Post Slack message"
  id: slack-post
  uses: ynab/slack-post-message-action@db/better-logging
  with:
    channel: "#my-channel"
    text: "Test 123"
- run: echo ${{ steps.slack-post.outputs.ts }}
```

dbertram replied:
The bigger issue here was that I needed to define how the `ts` output gets its value. The `Invalid format` aspect _did_ need to be fixed, but didn't actually correct the problem I was debugging: the `ts` output not being available in later steps for actions using this shared action.

dbertram commented on action.yml:
```diff
@@ -23,11 +27,17 @@ runs:
+          --header "Content-Type: application/json; charset=utf-8" \
+          --url https://slack.com/api/chat.postMessage \
+          --data "{\"channel\": \"${{ inputs.channel }}\", \"thread_ts\": \"${{ inputs.thread_ts }}\", \"unfurl_links\": ${{ inputs.unfurl_links }}, \"text\": \"${{ inputs.text }}\"}" \
+        )
```
My goals with these changes are:
- make this step fail if we get a response with HTTP status >= 400 (i.e., didn't work)
- capture the response and log it (to make debugging easier if anything _does_ go wrong)
- set the `ts` output for this action by processing the curl response

But I'm a total bash/curl newb, so please feel free to make suggestions! 🙏

bradymholt commented on action.yml:
```diff
@@ -13,6 +13,10 @@ inputs:
+  unfurl_links:
+    description: "Whether links in the message should be unfurled"
+    type: boolean
+    default: true
```
Does this default to true in Slack API if not provided?

dbertram replied:
Their docs don't actually say what it defaults to, but my anecdotal testing seems to indicate yes

bradymholt commented on action.yml:
```diff
@@ -23,11 +27,17 @@ runs:
+          echo "$response" | jq --raw-output '.ts'
           echo "EOF"
         } >> "$GITHUB_OUTPUT"
       shell: bash
```
It might be nice to put `shell: bash` at top to get it out of the way of this long(er) script.

```
runs:
  using: 'composite'
  steps:
    - shell:bash
      run: |
         ...
```

dbertram replied:
https://github.com/ynab/slack-post-message-action/pull/3/commits/1f4882eb92b6ee6cc14e964c0f1e736d61a9bb2d

dbertram replied:
https://github.com/ynab/slack-post-message-action/pull/3/commits/1f4882eb92b6ee6cc14e964c0f1e736d61a9bb2d

bradymholt reviewed (approved):
Looks great.  I posted a few comments but nothing to hold this up.

dbertram committed:
move 'shell: bash' to top of step

dbertram committed:
action: validate inputs, single quotes for JSON, set ts output correctly

dbertram committed:
GH Workflow: add branch-bulider.yml to test changes before merge

- test sending a Slack message with all optional inputs using their default values
- test sending a Slack message with all optional inputs provided (reply message + link unfurling disabled)

dbertram commented on action.yml:
```diff
@@ -13,21 +13,37 @@ inputs:
 outputs:
   ts:
     description: "The timestamp ID of the message that was just posted"
+    value: ${{steps.slack-post-message.outputs.ts}}
```
Oof. That took a lot more iteration than I care to admit, but this was the bit I was missing: to set your custom action's output you have to define where it should pull the value from. 🤦

bradymholt replied:
😅

dbertram commented on action.yml:
```diff
@@ -13,21 +13,37 @@ inputs:
       shell: bash
+      run: |
+        [[ "${{ inputs.channel }}" ]] || { echo "required input 'channel' not specified" ; exit 1; }
+        [[ "${{ inputs.text }}" ]] || { echo "required input 'text' not specified" ; exit 1; }
```
Not strictly what I set out to cover here, but it felt reasonable to validate the required inputs explicitly to (hopefully) provide a better troubleshooting experience.

bradymholt replied:
👍
